keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,874 @@
+ 	// apiservers, this time will be quantized down to the nearest fifteen minutes.
+ 	LastContacted *time.Time `json:"lastContacted" protobuf:"bytes,opt,name=lastContacted"`
+}
+
+// APIResource specifies the name of a resource and whether it is namespaced.
+type APIResource struct {
+	// name is the plural name of the resource.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// singularName is the singular name of the resource.  This allows clients to handle plural and singular opaquely.
+	// The singularName is more correct for reporting status on a single item and both singular and plural are allowed
+	// from the kubectl CLI interface.
+	SingularName string `json:"singularName" protobuf:"bytes,6,opt,name=singularName"`
+	// namespaced indicates if a resource is namespaced or not.
+	Namespaced bool `json:"namespaced" protobuf:"varint,2,opt,name=namespaced"`
+	// group is the preferred group of the resource.  Empty implies the group of the containing resource list.
+	// For subresources, this may have a different value, for example: Scale".
+	Group string `json:"group,omitempty" protobuf:"bytes,8,opt,name=group"`
+	// version is the preferred version of the resource.  Empty implies the version of the containing resource list
+	// For subresources, this may have a different value, for example: v1 (while inside a v1beta1 version of the core resource's group)".
+	Version string `json:"version,omitempty" protobuf:"bytes,9,opt,name=version"`
+	// kind is the kind for the resource (e.g. 'Foo' is the kind for a resource 'foo')
+	Kind string `json:"kind" protobuf:"bytes,3,opt,name=kind"`
+	// verbs is a list of supported kube verbs (this includes get, list, watch, create,
+	// update, patch, delete, deletecollection, and proxy)
+	Verbs Verbs `json:"verbs" protobuf:"bytes,4,opt,name=verbs"`
+	// shortNames is a list of suggested short names of the resource.
+	ShortNames []string `json:"shortNames,omitempty" protobuf:"bytes,5,rep,name=shortNames"`
+	// categories is a list of the grouped resources this resource belongs to (e.g. 'all')
+	Categories []string `json:"categories,omitempty" protobuf:"bytes,7,rep,name=categories"`
+
+ 	// StorageVersionHash will be removed in the new Discovery API
+	// StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
+}
+```
+
+### Aggregation
+
+For the aggregation layer on the server, a new controller will be
+created to aggregate discovery for built-in types, apiextensions types
+(CRDs), and types from aggregated api servers.
+
+A post start hook will be added and the kube-apiserver health check
+should only pass if the discovery document is ready. Since aggregated
+api servers may take longer to respond and we do not want to delay
+cluster startup, the health check will only block on the local api
+servers (built-ins and CRDs) to have their discovery ready. For api
+servers that have not been aggregated, their group-versions will be
```

Comment:
Is this a typo?  "have not been aggregated"?  I expected to see a description of what client visible behavior we'd see for the discovery doc on initialization:

1. Built-ins - always visible?
2. CRDs - only visible once CRDs are loaded, so might be visible?
3. Aggregated APIs - could take materially longer to be visible?

I'd like to see some examples added to the KEP on the outcome for clients to clarify what they see during startup and how they know discovery is incomplete.  Will an incomplete discovery doc return a non 200 error message?  Will clients be expected to retry if the document is not complete?